apiserver/logsink.go

```diff
@@ -7,27 +7,105 @@ import (
 	"io"
 	"net/http"
 	"strings"
+	"sync"
 	"time"
 
 	"github.com/juju/errors"
 	"github.com/juju/loggo"
+	"github.com/juju/utils/clock"
 	"github.com/juju/version"
 	"gopkg.in/juju/names.v2"
 
 	"github.com/juju/juju/apiserver/logsink"
 	"github.com/juju/juju/apiserver/params"
 	"github.com/juju/juju/state"
+	"github.com/juju/juju/state/logdb"
+)
+
+const (
+	// dbLoggerBufferSize is the capacity of the log buffer.
+	// When the buffer fills up, it will be flushed to the
+	// database.
+	dbLoggerBufferSize = 1024
```

Comment:
It's ~the same as what's used in the juju/txn pruning code. I did do some testing, and saw that in a small deployment the buffer was basically never reached, which means we'll generally wait 2s to flush. Yes, I think we'll want to expose a tuning knob, but I'll probably do that in a followup. I'll do some more analysis first.